### DISCUSSION TITLE
Map Datatypes for transfer to Oracle DB

### DISCUSSION THREAD

**Discussion by JohnnyFruitbasket** [sqlalchemy/sqlalchemy#11300] [Usage Questions] [answered]
So, I am struggling to figure this one out, hopefully somebody can offer some insight. For the transfer of the Pandas DF to an Oracle DB, I wanted to use the mapping feature
```python
from sqlalchemy import create_engine, types
self.df.to_sql(name=self.table_name.upper(), con=engine, if_exists='replace', index=False, dtypes=dtype_mapping)
```
for which I am not sure how to map the Oracle Datatypes properly. In one attempt, I tried something like that
```python
        dtype_mapping = {
            'int64': types.NUMERIC,
            'float64': types.FLOAT,
            'object': types.CHAR, 
            'datetime64[ns]': types.TIMESTAMP, 
        }
```
In all the attempts, the mapping seems to be ignored and all I get are CLOBs etc. in the DB. What am I doing wrong? Is something missing? Thanks in advance

**Comment by CaselIT** [comment]
Hi,

That seems correct, try looking at the pandas? docs/support forum for further help.
Not sure if you need to instantiate the sql types in the dict, so something like
```py
{
'int64': types.NUMERIC(),
'float64': types.FLOAT(),
'object': types.CHAR(), 
'datetime64[ns]': types.TIMESTAMP(), 
}

**Comment by gordthompson** [comment]
The pandas docs for to_sql()

https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_sql.html

only mention `dtype=`, not `dtypes=`, and the dict keys are column names.

**Comment by JohnnyFruitbasket** [comment]
Thanks for the input. Since the column names are created dynamically and I don't know how many columns the DF contains, I tried a more crude way:
```python
        for col in self.df.columns:
            if self.df[col].dtype == 'float64':
                self.df[col] = self.df[col].astype(dtype=types.FLOAT)
            if self.df[col].dtype == 'object':
                self.df[col] = self.df[col].astype(dtype=types.CHAR)
            if self.df[col].dtype == 'int64':
                self.df[col] = self.df[col].astype(dtype=types.NUMERIC)
```
But this attempt throws the following error: *TypeError: dtype '<class 'sqlalchemy.sql.sqltypes.CHAR'>' not understood*
Anybody knows what is happening there?

**Reply by CaselIT** [reply]
you can't convert pandas series to database types.

**Comment by JohnnyFruitbasket** [comment, marked as answer]
Ok, thanks for pointing me into the right direction. This seems to be more a Pandas issue. Anyhow, this workaround does it for me.
Create function:
```python
    def infer_dtype(self, series):
        if pd.api.types.is_numeric_dtype(series):
            return types.Float()  
        elif pd.api.types.is_string_dtype(series):
            return types.String(length=255)  
        elif pd.api.types.is_datetime64_any_dtype(series):
            return types.DateTime()
        # Default-Fallback
        else:
            return types.String(length=255)
```
Define dictionary and call function:
```python
dtype_dict = {col: self.infer_dtype(self.df[col]) for col in self.df.columns}
```
Use as dtype:
```python
self.df.to_sql(name=self.table_name.upper(), con=engine, if_exists='replace', index=False, dtype=dtype_dict)
```

**Reply by CaselIT** [reply]
looks like it makes sense.
Maybe you could have also a branch for integer values?

**Reply by JohnnyFruitbasket** [reply]
Thought about that, but there is no "is_integer_dtype", right? My main goal with the conversion was to eliminate the CLOB dataytpe in the Oracle DB, since these were causing problems with sql queries.

**Reply by CaselIT** [reply]
there seems to be https://pandas.pydata.org/docs/reference/api/pandas.api.types.is_integer_dtype.html
You likely need to have it as the first if branch, since is_numeric_dtype likely returns true also for integers

**Reply by JohnnyFruitbasket** [reply]
You are right. I added the option and it works like a charm.
